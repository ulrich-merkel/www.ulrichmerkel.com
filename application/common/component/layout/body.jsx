/** global picturefill */
/* eslint-disable immutable/no-mutation, immutable/no-this */
/**
 * Es6 module for React Component.
 * Layout components merge modules to bigger parts of the
 * overall page layout.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2016
 * @version 0.0.3
 *
 * @requires react
 * @requires prop-types
 * @requires react-helmet
 * @requires classnames
 * @requires common/component/decorator/picturefill
 * @requires common/component/decorator/scroller
 * @requires common/component/decorator/add-content
 * @requires common/utils/scroll-to
 * @requires common/state/constants
 * @requires common/component/layout/header
 * @requires common/component/layout/footer
 * @requires common/component/layout/loader
 * @requires common/component/layout/dialog
 * @requires common/component/page/broadcast
 * @requires common/component/page/search
 *
 * @changelog
 * - 0.0.3 Remove connect and state handling, improve dialog handling
 * - 0.0.2 Rewritten for es2015
 * - 0.0.1 Basic functions and structure
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Helmet from 'react-helmet';
import classnames from 'classnames';

import {
    pictureFill,
    scroller,
    addContent
} from '../decorator/';
import scrollTo, { getPageOffset } from '../../utils/scroll-to';
import {
    STATE_DIALOG_PAGE_BROADCAST,
    STATE_DIALOG_PAGE_SEARCH
} from '../../state/constants';
import {
    LayoutHeader,
    LayoutFooter,
    LayoutLoader,
    LayoutDialog
} from './index';
import PageBroadcast from '../page/broadcast';
import PageSearch from '../page/search';

/**
 * Class representing a component.
 *
 * @class
 * @extends React.Component
 * @property {Array|string} [props.children] - The component dom node childs - usally an array of components, if there is only a single child it's a string
 * @property {Object} [props.content={}] - The component content config
 */
class LayoutBody extends Component {

    /**
     * The actual class constructor.
     *
     * @constructs
     * @param {Object} [props] - The initial class properties
     * @returns {void}
     */
    constructor(props) {
        super(props);

        /**
         * Bind manually because React class components don't auto-bind.
         *
         * A bind call or arrow function in a JSX prop will create a brand new
         * function on every single render. This is bad for performance, as it
         * will result in the garbage collector being invoked way more than is necessary.
         *
         * Unfortunately React ES6 classes do not autobind their methods like components created
         * with the older React.createClass syntax. There are several approaches to binding methods
         * for ES6 classes. A basic approach is to just manually bind the methods in the constructor.
         *
         * @see {@link http://facebook.github.io/react/blog/2015/01/27/react-v0.13.0-beta-1.html#autobinding}
         * @see {@link http://stackoverflow.com/questions/23123138/perform-debounce-in-react-js/24679479#24679479}
         * @see {@link https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-no-bind.md}
         */
        this.handleScrollTop = this.handleScrollTop.bind(this);
    }

    /**
     * ShouldComponentUpdate is triggered before the re-rendering process starts,
     * giving the developer the ability to short circuit this process.
     *
     * @function
     * @param {Object} nextProps - The news props to be rendered
     * @returns {boolean} Whether to force component update or not
     */
    shouldComponentUpdate(nextProps) {
        return this.props !== nextProps;
    }

    /**
     * Scroll to top animation helper.
     *
     * @function
     * @param {Object} e - The synthetic react event
     * @returns {void}
     */
    handleScrollTop(e) { // eslint-disable-line class-methods-use-this
        e.preventDefault();
        if (getPageOffset()) {
            scrollTo({
                top: 0
            });
        }
    }

    /**
     * The required render function to return a single react child element.
     *
     * @function
     * @returns {React.Element} React component markup
     */
    render() {
        const {
            children,
            content
        } = this.props;

        const componentClassName = classnames('l-react__body');

        return (
            <div className={componentClassName}>
                <Helmet {...content} />
                <LayoutLoader />
                <LayoutHeader />
                {children}
                <LayoutFooter
                    handleScrollTop={this.handleScrollTop}
                />
                <LayoutDialog
                    page={STATE_DIALOG_PAGE_SEARCH}
                    isSearch
                >
                    <PageSearch isDialog />
                </LayoutDialog>
                <LayoutDialog
                    page={STATE_DIALOG_PAGE_BROADCAST}
                    isBroadcast
                >
                    <PageBroadcast isDialog />
                </LayoutDialog>
            </div>
        );
    }

}

/**
 * Validate props via React.PropTypes helpers.
 *
 * @static
 * @type {Object}
 */
LayoutBody.propTypes = {
    children: PropTypes.node, // eslint-disable-line react/require-default-props
    content: PropTypes.objectOf(PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
        PropTypes.array,
        PropTypes.object
    ]))
};

/**
 * Set defaults if props aren't available.
 *
 * @static
 * @type {Object}
 */
LayoutBody.defaultProps = {
    content: {}
};

/**
 * Connects a React component to a Redux store. It does not modify the
 * component class passed to it. Instead, it returns a new, connected component class.
 *
 * @type {React.Element}
 */
const LayoutBodyContainer = scroller(pictureFill(addContent('Head')(LayoutBody)));

export default LayoutBodyContainer;
export {
    LayoutBody
};
