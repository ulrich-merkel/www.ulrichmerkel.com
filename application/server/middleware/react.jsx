/**
 * Es6 module for react server middleware.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2016
 * @version 0.0.3
 *
 * @requires react
 * @requires react-dom
 * @requires react-router
 * @requires lodash
 * @requires common/config/application
 * @requires common/config/routes
 * @requires common/utils/logger
 * @requires common/utils/read-file
 * @requires common/component/root
 * @requires common/component/layout/html
 * @requires common/component/page/not-found
 * @requires common/state/configure-store
 * @requires common/state/config/actions
 * @requires common/state/intl/actions
 * @requires common/state/csrf/actions
 *
 * @see {@link https://github.com/reactjs/react-router/blob/1.0.x/docs/guides/advanced/ServerRendering.md}
 * @see {@link https://github.com/reactjs/react-router/issues/1990}
 * @see {@link https://www.youtube.com/watch?v=PnpfGy7q96U}
 * @see {@link https://github.com/reactjs/redux/issues/723}
 *
 * @changelog
 * - 0.0.3 Adjusted async rendering
 * - 0.0.2 Moved code to es6
 * - 0.0.1 Basic functions and structure
 */
import React from 'react';
import { renderToStaticMarkup } from 'react-dom/server';
import { RouterContext, match } from 'react-router';
import { get } from 'lodash';

import configApplication from './../../common/config/application';
import configRoutes from './../../common/config/routes';
import logger from './../../common/utils/logger';
import { readFile } from './../../common/utils/read-file';
import Root from './../../common/component/root';
import LayoutHtml from './../../common/component/layout/html';
import PageNotFound from './../../common/component/page/not-found';
import configureStore from './../../common/state/configure-store';
import { changeLocale } from './../../common/state/intl/actions';
import { fetchConfigContentIfNeeded, fetchConfigTranslationIfNeeded } from './../../common/state/config/actions';
import { addToken } from './../../common/state/csrf/actions';

const { aboveTheFold } = configApplication;

/**
 * Helper function to pass props to children components.
 *
 * @function
 * @private
 * @param {Object} Component The current router react component
 * @param {Object} props The current react component props
 * @returns {React.Element} React component markup
 */
function createElement(Component, props) {
    return <Component {...props} />;
}

/**
 * Using createElement function here to be able to
 * pass props to the page Component
 *
 * @see {@link https://github.com/reactjs/react-router/issues/3183}
 * @see {@link https://github.com/voronianski/universal-react-router-flux-2016/blob/master/src/server/middleware/renderHTML.js#L17}
 * @see {@link https://github.com/ReactTraining/react-router/issues/902}
 *
 * @function
 * @private
 * @param {Object} store The created redux store
 * @param {Object} renderProps The component properties to be rendered
 * @param {string} [cssBase=''] The file contents from base.css
 * @param {string} [scriptLoader=''] The file contents from loader.js
 * @returns {void}
 */
function getHtml(store, renderProps, cssBase = '', scriptLoader = '') {
    return renderToStaticMarkup(
        <Root store={store}>
            <LayoutHtml {... { store, cssBase, scriptLoader }}>
                <RouterContext
                    {...renderProps}
                    createElement={createElement}
                />
            </LayoutHtml>
        </Root>
    );
}

/**
 * Handle react server rendering and react routering.
 *
 * @function
 * @param {Object} req The current request object
 * @param {Object} res The result object
 * @returns {void}
 */
function middlewareReact(req, res, next) {

    /**
     * Using the new match function instead of Router.run
     *
     * Note that req.url here should be the full URL path from
     * the original request, including the query string.
     *
     * @see {@link https://github.com/reactjs/react-router/blob/1.0.x/docs/guides/advanced/ServerRendering.md}
     * @see {@link https://github.com/reactjs/react-router/blob/master/docs/guides/ServerRendering.md}
     */
    match({
        routes: configRoutes,
        location: req.url
    }, (error, redirectLocation, renderProps) => {

        if (error) {
            return res.status(error.status || 500).send(error.message).end();
        }

        if (redirectLocation) {
            return res.redirect(res.status || 302, `${redirectLocation.pathname}${redirectLocation.search}`);
        }

        if (renderProps) {

            // Initialize redux store, combining single stores into a global one
            const store = configureStore();

            /**
             * Read language which is given in the url locale parameter or via
             * express middleware and store the selection in redux
             */
            const urlLocale = get(renderProps, 'router.params.locale', '');
            store.dispatch(changeLocale([urlLocale, urlLocale.toUpperCase()].join('-'), req.language));
            const acceptedLocale = get(store.getState(), 'intl.locale');

            /**
             * We check renderProps.components (or renderProps.routes) for
             * the "not found" component or route respectively, and send a 404 as
             * below, because we're using a catch-all error route in the router
             * config.
             *
             * @see {@link https://github.com/ReactTraining/react-router/issues/2834}
             */
            const statusCode = get(renderProps, 'components', []).indexOf(PageNotFound) !== -1 ? 404 : 200;

            /**
             * Load all required data async via promises to
             * improve the overall node performance.
             */
            return Promise.all([
                readFile(aboveTheFold.baseCss),
                readFile(aboveTheFold.loaderBundle),
                store.dispatch(fetchConfigContentIfNeeded()),
                store.dispatch(fetchConfigTranslationIfNeeded(acceptedLocale)),
                store.dispatch(addToken(req.csrfToken && req.csrfToken()))
            ]).then((result) => {
                return res
                    .status(statusCode)
                    .send(`<!doctype html>${getHtml(store, renderProps, result[0], result[1])}`);
            }).catch((reason) => {
                logger.warn(reason);
                return res
                    .status(statusCode)
                    .send(`<!doctype html>${getHtml(store, renderProps)}`);
            });

        }

        return next();

    });
}

export default middlewareReact;
