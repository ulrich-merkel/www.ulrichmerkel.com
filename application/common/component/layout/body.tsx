/** global picturefill */
/**
 * Es6 module for React Component.
 * Layout components merge modules to bigger parts of the
 * overall page layout.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { default as React, PureComponent, ReactNode } from 'react';
import Helmet from 'react-helmet';
import { connect } from 'react-redux';

import { selectStateReducedMotionSelectedReduce } from '../../state/reduced-motion/selector';
import { pictureFill } from '../decorator/picturefill';
import { scroller } from '../decorator/scroller';
import { addContent } from '../decorator/add-content';
import { getPageOffset, scrollTo } from '../../utils/scroll-to';
import { eventPreventDefault } from '../../utils/event';
import {
    DIALOG_CONTENT_BROADCAST,
    DIALOG_CONTENT_SEARCH,
    DIALOG_CONTENT_THEME
} from '../../state/dialog/duck';
import { LayoutHeaderConnected } from './header';
import { LayoutFooterConnected } from './footer';
import { LayoutSettingsConnected } from './settings';
import { LayoutLoaderConnected } from './loader';
import { LayoutDialogConnected } from './dialog';
import { PageBroadcast } from '../page/broadcast';
import { PageSearch } from '../page/search';
import { PageSettings } from '../page/settings';

type Props = {
    children?: ReactNode;
    content?: any;
    reducedMotionSelectedReduce?: boolean;
};

/**
 * Class representing a component.
 *
 * @augments React.Component
 * @property {Array|string} [props.children] - The component dom node childs - usally an array of components, if there is only a single child it's a string
 * @property {object} [props.content={}] - The component content config
 */
export class LayoutBody extends PureComponent<Props> {
    /**
     * Scroll to top animation helper.
     *
     * @param {object} event - The synthetic react event
     * @returns {void}
     */
    handleScrollTop = (event) => {
        const { reducedMotionSelectedReduce = false } = this.props;

        eventPreventDefault(event);
        if (getPageOffset()) {
            scrollTo({
                duration: reducedMotionSelectedReduce ? 0 : 300,
                top: 0
            });
        }
    };

    /**
     * The required render function to return a single react child element.
     *
     * @returns {ReactElement} React component markup
     */
    render() {
        const { children, content } = this.props;

        return (
            <LayoutSettingsConnected>
                <Helmet {...content} />
                <LayoutLoaderConnected />
                <LayoutHeaderConnected />
                {children}
                <LayoutFooterConnected handleScrollTop={this.handleScrollTop} />
                <LayoutDialogConnected
                    className="l-dialog--search"
                    page={DIALOG_CONTENT_SEARCH}
                >
                    <PageSearch isDialog />
                </LayoutDialogConnected>
                <LayoutDialogConnected
                    className="l-dialog--broadcast"
                    page={DIALOG_CONTENT_BROADCAST}
                >
                    <PageBroadcast isDialog />
                </LayoutDialogConnected>
                <LayoutDialogConnected
                    className="l-dialog--theme"
                    page={DIALOG_CONTENT_THEME}
                >
                    <PageSettings isDialog />
                </LayoutDialogConnected>
            </LayoutSettingsConnected>
        );
    }
}

/**
 * The component will subscribe to Redux store updates. Any time it updates,
 * mapStateToProps will be called, Its result must be a plain object,
 * and it will be merged into the componentâ€™s props.
 *
 * @private
 * @param {object<string, *>} state - The current redux store state
 * @returns {object<string, *>} The mapped state properties
 */
function mapStateToProps(state) {
    return {
        reducedMotionSelectedReduce: selectStateReducedMotionSelectedReduce(
            state
        )
    };
}

/**
 * Connects a React component to a Redux store. It does not modify the
 * component class passed to it. Instead, it returns a new, connected component class.
 *
 * @type {ReactElement}
 */
export const LayoutBodyConnected = scroller(
    pictureFill(addContent('Head')(connect(mapStateToProps)(LayoutBody)))
);
