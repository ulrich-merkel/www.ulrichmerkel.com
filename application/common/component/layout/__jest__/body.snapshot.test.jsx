/* eslint-disable import/no-extraneous-dependencies, func-names, import/first */
import * as React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router';
import renderer from 'react-test-renderer';

import mockedStore from '../../__mocks__/store';
import { LayoutBody } from '../body';

describe('common/component/layout/body', function () {
    const defaultProps = {
        content: {}
    };

    it('should render correctly', function () {
        const tree = renderer
            .create(
                <Provider store={mockedStore}>
                    <MemoryRouter>
                        <LayoutBody {...defaultProps}>Body Children</LayoutBody>
                    </MemoryRouter>
                </Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
