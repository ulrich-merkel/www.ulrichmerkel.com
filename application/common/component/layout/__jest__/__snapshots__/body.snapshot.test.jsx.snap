// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/component/layout/body should render correctly 1`] = `
<div
  className="l-react__body"
>
  <dialog
    className="l-dialog--loading"
  >
    <div
      className="l-dialog__content"
    >
      <div
        className="c-flip"
      >
        <div
          className="c-flip__front"
        />
        <div
          className="c-flip__back"
        />
      </div>
    </div>
    <div
      className="l-dialog__background"
    />
  </dialog>
  <header
    className="l-header is-visible"
    itemScope={true}
    itemType="http://schema.org/WPHeader"
    role="banner"
  >
    <progress
      className="m-progress"
      id="m-progress"
      max="100"
      value="0"
    >
      <span
        className="m-progress__fallback"
        id="m-progress__fallback"
      />
    </progress>
    <nav
      className="m-nav m-nav--main"
      role="navigation"
    >
      <section
        className="l-grid"
      >
        <div
          className="l-grid__row"
        >
          <div
            className="l-grid__col--12"
          >
            <input
              aria-label={undefined}
              className="m-nav__toggle c-font-icon--menu"
              data-label={undefined}
              type="checkbox"
            />
            <ul
              className="m-menu--aside m-nav__toggle-target m-nav__aside"
              itemScope={true}
              itemType="http://schema.org/ItemList"
              role="menu"
            >
              <li
                className="m-menu__list-item"
                itemProp="itemListElement"
                itemScope={true}
                itemType="http://www.schema.org/SiteNavigationElement"
              >
                <button
                  className="c-btn m-menu__item m-menu__item--en is-active is-hidden c-btn--small c-btn--clear"
                  data-locale="en-EN"
                  onClick={[Function]}
                  role="button"
                  title=""
                  to={undefined}
                  type="button"
                >
                  <span
                    className="c-btn__label"
                  />
                </button>
              </li>
              <li
                className="m-menu__list-item"
                itemProp="itemListElement"
                itemScope={true}
                itemType="http://www.schema.org/SiteNavigationElement"
              >
                <button
                  className="c-btn m-menu__item m-menu__item--de is-hidden c-btn--small c-btn--clear"
                  data-locale="de-DE"
                  onClick={[Function]}
                  role="button"
                  title=""
                  to={undefined}
                  type="button"
                >
                  <span
                    className="c-btn__label"
                  />
                </button>
              </li>
              <li
                className="m-menu__list-item"
                itemProp="itemListElement"
                itemScope={true}
                itemType="http://www.schema.org/SiteNavigationElement"
              >
                <a
                  aria-current={false}
                  className="c-link m-menu__item--search c-btn--small c-btn--clear"
                  href="/search"
                  onClick={[Function]}
                  style={undefined}
                  tabIndex="0"
                  title="Suche"
                >
                  <span
                    className="c-btn__label"
                  >
                    <i
                      aria-hidden="true"
                      className="c-font-icon--search c-btn__icon"
                    />
                  </span>
                </a>
              </li>
            </ul>
          </div>
        </div>
      </section>
    </nav>
  </header>
  Body Children
  <footer
    className="l-footer"
    itemScope={true}
    itemType="http://schema.org/WPFooter"
    role="contentinfo"
  >
    <button
      className="c-btn l-footer__button--up c-font-icon--chevron-up"
      onClick={[Function]}
      role="button"
      title=""
      to={undefined}
      type="button"
    >
      <span
        className="c-btn__label is-visually-hidden"
      />
    </button>
    <nav
      className="m-nav"
      role="navigation"
    >
      <section
        className="l-grid"
      >
        <div
          className="l-grid__spaced--small"
        >
          <div
            className="l-grid__row"
          >
            <div
              className="l-grid__col--12"
            />
          </div>
          <div
            className="l-grid__row"
          >
            <div
              className="l-grid__col--12"
            />
          </div>
          <div
            className="l-grid__row"
          >
            <div
              className="l-grid__col--12"
            >
              <small
                className="c-type--small"
              >
                Â© 
                <strong />
                 
              </small>
            </div>
          </div>
        </div>
      </section>
    </nav>
  </footer>
</div>
`;
